### PR TITLE
Update primeline so that it runs along the rectangle of the first layer

### PR DESCRIPTION
**Pull request by elpopo-eng**
Thanks to @fbeauKmi 

Now `prime_line_xy` defines the origin of the baseline
The primeline is made on the corner of the `FL_SIZE` rectangle closest to the origin.

Add Variable `prime_line_margin` wich define distance of purge line from fl_size rectangle

If `FL_SIZE` is not defined, old primeline is used

### CODE REVIEW

**Comment by Surion79** [comment]
i don't see the benefit of it except 2 seconds of travel. for me it looks overcomplicating stuff, including possible debugging.

Aside of that: This will fail after updating klippain via moonraker as it is defaulting on a non-existing variable.

**Comment by elpopo-eng** [comment]
The aim is to place the primeline as close as possible to the bdemesh area.
If we place the primeline on the edge of the bed, you can engrave your PEI, but if you like that, we can introduce a variable to use the legacy baseline...

**Comment by Surion79** [comment]
Engrave? The location is so near the border, it would never get printed on anyway.

**Comment by Surion79** [comment]
Don't forget my other comments regarding breaking klippain on update without the default fix.

File: macros/helpers/prime_line.cfg
Comment (Surion79) [inline macros/helpers/prime_line.cfg:36]: This will break klippain on moonraker update if a fixed value default isn't added.
We had that topic on mmu already.

File: macros/helpers/prime_line.cfg
Comment (Surion79) [inline macros/helpers/prime_line.cfg:39]: This verifies that x/y never goes under 0/min or over max?

**Review by Surion79 — DISMISSED** [review]
The default on a non existing variable will break *all* installations which are updating existing klippain installations. Fix is to add another default with the default value. That was the solution @Benoitone suggested when we worked on the mmu verification of his branch.

**Comment by fbeauKmi** [comment]
Thanks for this PR @elpopo-eng . I think this is a nice to have. IMO the purge line should always be inside the bed mesh.Then it could be possible to go lower at the end of the prime line (Z<0.05) to wipe the nozzle on the bed.  

@Surion79 If you don't intend to use this feature, you can remove prime_line from your startup sequence and then add it to your startup gcode in the slicer.

That's the beauty of Klippain being fully overridable ❤️ .

File: macros/helpers/prime_line.cfg
Comment (fbeauKmi) [reply macros/helpers/prime_line.cfg:40]: This ensure that prime line start point is inside FL_SIZE area + prime_line_margin.
The prime_line_way parameter attempt to avoid that prime_line leaves the print area in case of misconfiguration of prime line start point.

**Comment by Surion79** [comment]
> Thanks for this PR @elpopo-eng . I think this is a nice to have. IMO the purge line should always be inside the bed mesh.Then it could be possible to go lower at the end of the prime line (Z<0.05) to wipe the nozzle on the bed.
> 
> @Surion79 If you don't intend to use this feature, you can remove prime_line from your startup sequence and then add it to your startup gcode in the slicer.
> 
> That's the beauty of Klippain being fully overridable ❤️ .

errr.... My Klippain tag is reading Collaborator, and i proposed a solution? I could also just push my proposal to the Klippain branch `primeline_FLSIZE_dependent`, but I won't.

Anyway, you don't override stuff by making users override existing functionality in klippain to tell them "do that in your slicer now".

**Comment by fbeauKmi** [comment]
> Anyway, you don't override stuff by making users override existing functionality in klippain to tell them "do that in your slicer now".

No offense, override in Klippain if you prefer 👍

**Comment by Surion79** [comment]
still a no regarding overrides, because this change contradicts existing configuration in variables.cfg. without proper parametrization or setting an enable/disable variable, the prime line location is moved away from the settings in the variables.cfg.
i just ask for a variable to disable/enable this feature. i have no idea, why this is such hassle? it is 2 lines of code and a variable, additional to the variable which is already introduced by this PR.

**Comment by Frix-x** [comment]
I appreciate the discussion surrounding this PR and the thoughtful contributions from everyone involved, but please keep it friendly. To address the concerns raised, here is my point:

1. The demand for adaptive purge integrated within Klippain is something that I get everyday on Discord or in PMs. As currently there is no solution in Klippain, I push the users over using KAMP. The problem is that using KAMP in Klippain is not that easy and can easily break the adaptive mesh: this create a lot of support work for me to help them going. So I'm definitely full on to integrate this to Klippain to have something just working out of the box.

1. The initiative for developing adaptive purge has been collaboratively discussed on Discord and is supported by contributions from Elpopo and Fbeaukmi since I had no time to develop it myself. This collaboration allowed to bring this PR and is also followed by a lot of people on Discord just waiting for it to be merged.

1. The concerns about a printed line (as the purge operation is doing) outside the meshed print zone are valid, particularly for printers with warped beds like Ender3, Prusa or Switchwire. This was mitigated by using a fixed height of 0.6mm for the prime line, but it was changed in a previous commit with an adjustable height parameter, that, while enhancing flexibility, also reintroduces the risk of crashes if using a small value closer to the print layer height like 0.2mm. Positioning the purge as close as possible to the part emerges as a prudent strategy to mitigate these risks.

1. Finally, I think that having adaptive mesh enabled by default sets a precedent for incorporating adaptive purge in a similar manner. Users can still not use the option by removing the SIZE parameter. I do not see any point in doing an adaptive mesh and not an adaptive purge or vice versa. On a good bed, it will be nothing more than a shiny feature but will not have any cons, on a warped bed it can avoid a crash.

This PR represents a step forward in aligning Klippain's features with the community's evolving needs and Klipper's advancements (like the adaptive mesh that is now mainline). So I'll review and test it properly to validate that everything is working and merge it in the current state.

**Comment by Surion79** [comment]
I still request to add a disable variable, so users can choose to make a bed mesh and an outside prime line.
It can be enabled by default.

@Frix-x  Since i invested a lot of time to this project, i think it is not too much to ask.

**Comment by elpopo-eng** [comment]
@Frix-x we have chosen to always orient the primeline towards the centre of the bed (case 1) in order to avoid a possible end of prime_line outside the bed (case 2) and to have a simple code.
We can change this behaviour if you don't think it's correct.

case 1 : 
image

case 2 :
image

**Comment by Frix-x** [comment]
Ok, I made some changes to make it a bit more flexible and cleaner:
  - There is now an `ADAPTIVE_PRIMELINE` variable that you can use from your slicer to enable/disable the "adaptive mode" of the primeline directly from the START_PRINT sequence. By default, if nothing is set, it's True, so adaptive mode will be enabled for everyone unless we force it to be disabled.
  - The `PRIMELINE' macro can now also take an `ADAPTIVE_MODE' parameter that allow you to set the behavior when calling it manually. I think this is useful because when called manually, there is no printing going on and I understand that someone would want to know where the primeline is going to be and not randomly put it where the last print was done.
  - I've extended the SIZE parameter to work as it does with the bed_mesh by also allowing the exclude objects tags to be used when they are found and when SIZE is omitted. So in this case it will default to something like KAMP.
  - I also added a very small move at the end to prevent the head from crashing back in the primeline if the slicer starts to emit a Z down move, as I noticed some actually do and this was asked multiple times by the community.

As my main printer is currently running for a couple of hours, can someone test it and let me know if everything is working ok?

**Comment by Surion79** [comment]
I habe currently an open printer and don't mind testing it. Anything specific to test? Fl size near the border?

File: macros/helpers/prime_line.cfg
Comment (fbeauKmi) [inline macros/helpers/prime_line.cfg:15]: IMHO, using exclude_object to define the prime line introduces risks of collision. Although there's a lot of work being done in this area, not all slicers use exclude_object for brim, skirts and purge towers. 

But if you feel that this code should be maintained :
This code is already used in Adaptive_bed_mesh, would it be interesting to take it out of these 2 macros to introduce a mechanism to populate FL_SIZE whith exclude object during START_PRINT, for example or as reverse define an exclude_object with FL_SIZE ? Having in mind the compatibility with Klippe adaptive_mesh.

File: macros/helpers/prime_line.cfg
Comment (fbeauKmi) [inline macros/helpers/prime_line.cfg:43]: Can we put the params.LINE_MARGIN line with other entry params for readability ?

As discuss with @elpopo-eng in https://github.com/Frix-x/klippain/pull/498#issuecomment-1963513548 there is an edge case where primeline could be outside the mesh. 
It add a bit of complixity but I suggest this code to mitigate this situation : 
```
# We find bed center : useful to determine pime_line position and prime_line_way
    {% set center_x, center_y  = [printer.toolhead.axis_maximum.x / 2, printer.toolhead.axis_maximum.y / 2]|map("float")  %}

    # If first layer is retrieved and adaptive mode is enabled, then we replace the coordinates to do an adaptive purge
    {% if if coordinatesFound and prime_line_adaptive" %}
        {% set min_x, min_y, max_x, max_y = fl_size.split("_")|map('float') %}

        # Fix prime_line start position while FL_SIZE is not in the same bed area
        {% set prime_line_x = 2*center_x - prime_line_x 
                               if (prime_line_x > center_x and xMaxSpec < center_x) or (prime_line_x < center_x and xMinSpec > center_x) 
                               else prime_line_x %}
        {% set prime_line_y = 2*center_y - prime_line_y 
                               if (prime_line_y > center_y and yMaxSpec < center_y) or (prime_line_y < center_y and yMinSpec > center_y) 
                               else prime_line_y %}

    
        {% set prime_line_x = [[prime_line_x, xMinSpec - prime_line_margin] | max, xMaxSpec + prime_line_margin] | min %}
        {% set prime_line_y = [[prime_line_y, yMinSpec - prime_line_margin] | max, yMaxSpec + prime_line_margin] | min %}
    {% endif %}
    
    # Choose the way of printing the primeline (in + or -) alongside the direction to avoid going outside the bed boundaries
    {% set prime_line_way = -1 if (prime_line_direction == "X" and prime_line_x > center_x) or (prime_line_direction == "Y" and prime_line_y > center_y) else 1 %}
```

File: macros/helpers/prime_line.cfg
Comment (fbeauKmi) [inline macros/helpers/prime_line.cfg:77]: The error is raised in Verbose mode only , is it mandatory ?

File: macros/helpers/prime_line.cfg
Comment (Frix-x) [reply macros/helpers/prime_line.cfg:15]: This will be reworked later as I do plan to switch to the mainline adaptive bed mesh at some point when slicer will manage correctly the purge tower as an object. And I'll indeed work on defining fake objects definitions to be able to use it the same way in manual calibrations macros and make it transparent for an user.

File: macros/base/start_print.cfg
Comment (Frix-x) [reply macros/base/start_print.cfg:32]: You are right, I'll use 0/1 since it come from the slicer and will be easier to manage this way!

File: macros/helpers/prime_line.cfg
Comment (Frix-x) [reply macros/helpers/prime_line.cfg]: I fixed it using your code, thanks!

**Comment by Frix-x** [comment]
New push, new tests needed before a final merge :)

**Pull request merged**
